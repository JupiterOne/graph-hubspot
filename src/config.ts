import {
  IntegrationExecutionContext,
  IntegrationInstanceConfig,
  IntegrationInstanceConfigFieldMap,
  IntegrationValidationError,
} from '@jupiterone/integration-sdk-core';
import { createAPIClient } from './client';

export const instanceConfigFields: IntegrationInstanceConfigFieldMap = {
  oauthAccessToken: {
    type: 'string',
    mask: true,
  },
  apiBaseUrl: {
    type: 'string',
    mask: true,
  },
};

/**
 * Properties provided by the `IntegrationInstance.config`. This reflects the
 * same properties defined by `instanceConfigFields`.
 */
export interface IntegrationConfig extends IntegrationInstanceConfig {
  /**
   * This access_token is considered valid as a Bearer Authorization header
   */
  oauthAccessToken: string;

  /**
   * Hubspot API base url
   */
  apiBaseUrl: string;
}

export async function validateInvocation(
  context: IntegrationExecutionContext<IntegrationConfig>,
) {
  const { config } = context.instance;
  if (!Object.keys(instanceConfigFields).every((key) => config[key])) {
    throw new IntegrationValidationError(
      `Config requires all of {${Object.keys(instanceConfigFields).join(
        ', ',
      )}}`,
    );
  }

  const apiClient = createAPIClient(config, context.executionHistory);
  await apiClient.verifyAuthentication();
}
