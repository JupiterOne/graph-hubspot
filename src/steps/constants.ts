import {
  RelationshipClass,
  StepEntityMetadata,
  StepRelationshipMetadata,
} from '@jupiterone/integration-sdk-core';

export enum IntegrationSteps {
  OWNERS = 'fetch-owners',
  COMPANIES = 'fetch-companies',
  DOMAINS = 'fetch-domains',
  TEAMS = 'fetch-teams',
  CONTACTS = 'fetch-contacts',
}

export const Entities: Record<
  'USER' | 'TEAM' | 'CONTACT' | 'COMPANY' | 'DOMAIN',
  StepEntityMetadata
> = {
  /*
     Depending on the other resources later (e.g. will we encounter regular users?)
     We may want to call this "Owner" or similar, we can still use the same class of "User"
     Since user can be both a 'regular' user and the owner.
     A bit hard to tell for sure right now
  */
  USER: {
    _type: 'hubspot_user',
    _class: 'User',
    resourceName: 'HubSpot User',
  },

  TEAM: {
    _type: 'hubspot_team',
    _class: 'Team',
    resourceName: 'HubSpot Team',
  },

  CONTACT: {
    _type: 'hubspot_contact',
    _class: 'User',
    resourceName: 'Hubspot Contact',
  },

  COMPANY: {
    _type: 'hubspot_company',
    _class: 'Organization',
    resourceName: 'HubSpot Company',
  },

  DOMAIN: {
    _type: 'hubspot_domain',
    _class: 'Domain',
    resourceName: 'Hubspot domain',
  },
};

export const Relationships: Record<
  'USER_HAS_TEAM',
  StepRelationshipMetadata
> = {
  USER_HAS_TEAM: {
    _type: 'hubspot_user_has_team',
    _class: RelationshipClass.HAS,
    sourceType: Entities.USER._type,
    targetType: Entities.TEAM._type,
  },
};
