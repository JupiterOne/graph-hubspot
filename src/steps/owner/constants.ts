export const OwnerEntity = {
  RESOURCE_NAME: 'Owner',
  TYPE: 'acme_user',
  CLASS: 'User',
};
