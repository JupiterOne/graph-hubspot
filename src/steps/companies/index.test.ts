import { createMockStepExecutionContext } from '@jupiterone/integration-sdk-testing';
import * as company from '.';
import { createIntegrationConfig } from '../../../test/config';
import { IntegrationConfig } from '../../config';
import { Entities } from '../constants';

describe('#fetchCompanies', () => {
  // TODO: we'll add recordings later
  // let recording: Recording;

  // beforeEach(() => {
  //   recording = setupHubSpotRecording({
  //     directory: __dirname,
  //     name: 'fetchOwners',
  //   });
  // });

  // afterEach(async () => {
  //   await recording.stop();
  // });

  test('should collect data', async () => {
    const context = createMockStepExecutionContext<IntegrationConfig>({
      instanceConfig: createIntegrationConfig(),
    });

    await company.fetchCompanies(context);

    expect({
      numCollectedEntities: context.jobState.collectedEntities.length,
      numCollectedRelationships: context.jobState.collectedRelationships.length,
      collectedEntities: context.jobState.collectedEntities,
      collectedRelationships: context.jobState.collectedRelationships,
      encounteredTypes: context.jobState.encounteredTypes,
    }).toMatchSnapshot();

    const companies = context.jobState.collectedEntities.filter((e) =>
      e._class.includes(Entities.COMPANY._class as string),
    );

    expect(companies.length).toBeGreaterThan(0);
    expect(companies).toMatchGraphObjectSchema({
      _class: [Entities.COMPANY._class as string],
      schema: {
        additionalProperties: false,
        required: [],
        properties: {
          _rawData: {
            type: 'array',
            items: { type: 'object' },
          },
          _type: { const: Entities.COMPANY._type },
          // We want to make sure we include every field from converter.ts
          name: { type: 'string' },
          website: { type: 'string' },
          emailDomain: { type: 'string' },
          archived: { type: 'boolean' },
          createdOn: { type: 'number' },
          updatedOn: { type: 'number' },
        },
      },
    });
  });
});
