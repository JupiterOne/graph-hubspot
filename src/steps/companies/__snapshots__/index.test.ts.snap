// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#fetchCompanies should collect data 1`] = `
Object {
  "collectedEntities": Array [
    Object {
      "_class": Array [
        "Organization",
      ],
      "_key": "hubspot_company:6534491450 ",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": SimplePublicObjectWithAssociations {
            "archived": false,
            "archivedAt": undefined,
            "associations": undefined,
            "createdAt": 2021-07-08T05:09:01.918Z,
            "id": "6534491450",
            "properties": Object {
              "createdate": "2021-07-08T05:09:01.918Z",
              "domain": "hubspot.com",
              "hs_lastmodifieddate": "2021-07-08T05:10:03.397Z",
              "hs_object_id": "6534491450",
              "name": "HubSpot, Inc.",
            },
            "updatedAt": 2021-07-08T05:10:03.397Z,
          },
        },
      ],
      "_type": "hubspot_company",
      "archived": false,
      "city": "",
      "createdOn": 2021,
      "displayName": "HubSpot, Inc.",
      "domain": "hubspot.com",
      "host": "",
      "id": "6534491450",
      "industry": "",
      "name": "HubSpot, Inc.",
      "ownerId": "",
      "public": false,
      "updatedOn": 2021,
    },
    Object {
      "_class": Array [
        "Organization",
      ],
      "_key": "hubspot_company:6558739580 ",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": SimplePublicObjectWithAssociations {
            "archived": false,
            "archivedAt": undefined,
            "associations": undefined,
            "createdAt": 2021-07-13T05:34:55.981Z,
            "id": "6558739580",
            "properties": Object {
              "createdate": "2021-07-13T05:34:55.981Z",
              "domain": null,
              "hs_lastmodifieddate": "2021-07-13T05:35:58.142Z",
              "hs_object_id": "6558739580",
              "name": "Test company",
            },
            "updatedAt": 2021-07-13T05:35:58.142Z,
          },
        },
      ],
      "_type": "hubspot_company",
      "archived": false,
      "city": "",
      "createdOn": 2021,
      "displayName": "Test company",
      "domain": "",
      "host": "",
      "id": "6558739580",
      "industry": "",
      "name": "Test company",
      "ownerId": "",
      "public": false,
      "updatedOn": 2021,
    },
  ],
  "collectedRelationships": Array [],
  "encounteredTypes": Array [
    "hubspot_company",
  ],
  "numCollectedEntities": 2,
  "numCollectedRelationships": 0,
}
`;
